modules/mobile-agents/windows/jax-rs/src/main/java/org/wso2/carbon/mdm/mobileservices/windows/services/wstep/util/CertificateSigningService.java

```diff
@@ -0,0 +1,137 @@
+
+	private static final long MILLI_SECONDS = 1000L * 60 * 60 * 24;
+	public static final int FIRST_ITEM = 0;
+	public static final int SECOND_ITEM = 1;
+	public static final int THIRD_ITEM = 2;
+	private static Log logger = LogFactory.getLog(CertificateSigningService.class);
+
+	/**
+	 * Implement certificate signing task using CSR received from the device and the MDM server key
+	 * store.
+	 *
+	 * @param jcaRequest        - CSR from the device
+	 * @param privateKey        - Private key of CA certificate in MDM server
+	 * @param CACert            - CA certificate in MDM server
+	 * @param certParameterList - Parameter list for Signed certificate generation
+	 * @return - Signed certificate for CSR from device
+	 * @throws CertificateGenerationException
+	 * @throws XMLFileOperationException
+	 */
+	public static X509Certificate signCSR(JcaPKCS10CertificationRequest jcaRequest,
+	                                      PrivateKey privateKey, X509Certificate CACert,
+	                                      List certParameterList)
+			      throws CertificateGenerationException, XMLFileOperationException {
+
+		String commonName = (String) certParameterList.get(FIRST_ITEM);
+		int notBeforeDays = (Integer) certParameterList.get(SECOND_ITEM);
+		int notAfterDays = (Integer) certParameterList.get(THIRD_ITEM);
+		X509v3CertificateBuilder certificateBuilder;
+		X509Certificate signedCertificate;
+
+		try {
+			ContentSigner signer;
+			BigInteger serialNumber = BigInteger.valueOf(new SecureRandom().nextInt(Integer.MAX_VALUE));
+			Date notBeforeDate = new Date(System.currentTimeMillis() -
+			                             (MILLI_SECONDS * notBeforeDays));
+			Date notAfterDate = new Date(System.currentTimeMillis() +
+			                             (MILLI_SECONDS * notAfterDays));
+			certificateBuilder =
+				  new JcaX509v3CertificateBuilder(CACert, serialNumber, notBeforeDate, notAfterDate,
+				                                  new X500Principal(commonName),
+				                                  jcaRequest.getPublicKey());
+
+			//Adding extensions to the signed certificate.
+			certificateBuilder.addExtension(Extension.keyUsage, true,
+			                                new KeyUsage(KeyUsage.digitalSignature));
+			certificateBuilder.addExtension(Extension.extendedKeyUsage, false,
+			                                new ExtendedKeyUsage(KeyPurposeId.id_kp_clientAuth));
+			certificateBuilder.addExtension(Extension.basicConstraints, true,
+			                                new BasicConstraints(false));
+
+			signer = new JcaContentSignerBuilder(Constants.CertificateEnrolment.ALGORITHM).setProvider(
+					Constants.CertificateEnrolment.PROVIDER).build(privateKey);
+
+			signedCertificate = new JcaX509CertificateConverter().setProvider(
+					Constants.CertificateEnrolment.PROVIDER).getCertificate(
+					certificateBuilder.build(signer));
+		} catch (InvalidKeyException e) {
+			String msg = "CSR's public key is invalid";
```

Comment:
Log and throw is generally considered an anti-pattern. Just throwing the exceptions would be enough in instances like this.